keps/sig-apps/0012-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,162 @@
++	 // Absolute number is calculated from percentage by rounding down.
++	 BatchSize intstr.IntOrString
++ }
+```
+
+User specifies the new update strategy type in the spec and issue the request
+
+### Deployment Controller
+
+#### Update
+Below is the proposed approach for in-place update.
+Wait until expected number of pods created, when deployment controller receives the in-place update request
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    // BatchSize is the number of pods to update at each round.
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+
+    // swap oldReplicaSet and newReplicaSet information such as spec, revision, annotation etc.
+    copyInfo = oldReplicaSet // write into etcd, in case the controller failed that may cause oldReplicaSet lost
+    oldReplicaSet = newReplicaSet
+    newReplicaSet = copyInfo
+```
+In essence, this approach replaces the current ReplicaSet with the new ReplicaSet info. The pods are always managed by 
+the same ReplicaSet object, but just its content gets swapped.
+
+We propose not updating the creationTimeStamp of the ReplicaSet, since pods are not re-created and creation-timestamp also stays the same.
+
+#### Rollback
+Rollback will be supported in a similar fashion as the existing rollback strategy. It first checkout the old ReplicaSet and then do update.
+To do the update,  it'll use the same steps as above to update the pods in-place, instead of deleting old pods and creating new pods.
+
+### Alternatives
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+    
+    // Pause the old replica set. This is required for this scenario: if we reduce oldReplicaSet immediately, there's a chance
+    // that oldReplicaSet may kill the pods prematurely.
+    pause(oldReplicaSet)
+    
+    // Reduce oldReplicaSet by the BatchSize
+    scaleDown(oldReplicaSet, replicas - BatchSize)
+    
+    // update pod controllerRef to the new ReplicaSet
+    for pod range(podsToUpdate)
+        update(pod, newReplicaSet-controllerRef)
+
+    // Increase newReplicaSet by BatchSize
+    scaleUp(newReplicaSet, replicas + BatchSize)
+
+```
+This approach requires some changes in ReplicaSet controller to support pause semantics. Another way is to pass certain information into ReplicaSet and prevent it from prematurely killing
+the pods while updating the pods' controllerRef and ReplicaSet's replicas. Or make Kubernetes support transactions which is another big effort.
```

Comment:
Not sure how transactions would help. Mind expanding all alternatives into separate sections and elaborating as much as you can on each of them?

Reply:
Yes, add in the doc